### ISSUE TITLE
Extended support for C coverage measurements

### ISSUE THREAD

**Issue by ionelmc** [codecov/codecov-python#28]
The current integration is fairly inconvenient for large-ish projects: https://codecov.io/guide#c

Would be very nice if codecov would be able to run gcov and collect everything automatically, just like [cpp-coveralls](https://github.com/eddyxu/cpp-coveralls/blob/master/cpp_coveralls/coverage.py#L178-L270)

**Comment by stevepeak** [comment]
@ionelmc thank you for the feedback.

If I could ask you to try the following during your CI build it would be great: 
- simply remove the part about `gcov :file` 
- run the command below to upload to codecov

```yml
after_success:
  - bash <(curl -s https://raw.githubusercontent.com/codecov/codecov-bash/wip/codecov)
```

Here is my commit for this issue: https://github.com/codecov/codecov-bash/commit/54846063ba4e42df44b1d2ffc84cc90932df8bd5

Thanks!

**Comment by ionelmc** [comment]
@stevepeak Doesn't seem to match very well my sourcecode: https://codecov.io/github/ionelmc/python-lazy-object-proxy

Would you implement something like cpp-coveralls in codecov-python?

**Comment by ionelmc** [comment]
@stevepeak Or, would you accept such a change?

**Comment by stevepeak** [comment]
I would implement it in the Bash uploader. It does not require `sudo` which is a benefit for many customers. I can add it to the Python uploader as well. (The bash works for python as well). Did you try to new upload method rom above? Can you send me the CI build that it ran on.

The files seem to line up. Just that first one is extra. Which you can remove in our UI under **Features -> Ignore**

**Comment by ionelmc** [comment]
I just ran `bash <(curl -s https://raw.githubusercontent.com/codecov/codecov-bash/wip/codecov) -t d5e1ecf0-2ead-4233-81eb-cb336a9ef083` locally (use the same thing to run tests locally and on CI - `tox`)

**Comment by ionelmc** [comment]
@stevepeak the first file is matched to the wrong path. It should be matched to src/lazy_object_proxy/_proxy.c 

I don't know what cpp-coveralls does but it works :grin:

**Comment by stevepeak** [comment]
That is odd indeed. We have an internal function that will do its best at determining file paths. But in this case it has seemed to fail.

Could you fun this command and share the output with me:
```
bash <(curl -s https://raw.githubusercontent.com/codecov/codecov-bash/wip/codecov) -vd
```

I can review why the path does not line up to patch our services. Otherwise, I believe this package is working as expected and can be used for your project. Thank you!

**Comment by ionelmc** [comment]
Here's the output: http://paste.ubuntu.com/11670448/

It's beyond my control what happens with the build dir and how the .so files are moved around :disappointed:

**Comment by ionelmc** [comment]
I've used `cpp-coveralls --build-root=. --include=src` to fix this issue with relocation. Could you implement some options like those? (the code in cpp-coveralls is there for the taking :grin:)

Also, I'm a bit confused - what's the difference between this shell script and the codecov python package? Why/when should I use the later?

**Comment by stevepeak** [comment]
I'll go about adding the support as suggested from cpp-coveralls. I'm going to fork your project to play around with the configuration of the Bash uploader, which for now should be used.

----

TL;DR; For the most part the `bash` and `python` uploaders are the same. Originally the python uploader was used for many projects (just like the bash). The bash was created so that the need for `sudo pip install codecov` was not necessary and when using Travis you can safely upload without `sudo` required. The bash uploader also is a more streamline approach so that when I add new features or optimization its added to one repo vs dozens (for each language)

In sum: the `bash` uploader is our "global" uploader that works for most languages :+1: , including python. And is one line of code :+1:.

**Comment by ionelmc** [comment]
I think there's some confusion here, on Travis sudo is never required for
`pip install` as everything runs in a virtualenv.

Plus I'm a bit wary of anything implemented in Bash - the language is
riddled with issues and pitfalls.


Thanks,
-- Ionel Cristian Mărieș, http://blog.ionelmc.ro

**Comment by stevepeak** [comment]
You are correct about sudo not being required, but this is only true when the language is set to python. Otherwise sudo is required for every other language that the Bash uploader is used for. I'm not to concerned about the pitfalls of bash. The uploader is very straight forward and should not have issues doing the job well.

I'll still look into adding support for C in in this project. I don't know if I can [copy/paste the code that does this in cpp-coveralls](https://github.com/eddyxu/cpp-coveralls/blob/master/cpp_coveralls/coverage.py#L178-L223) because of licenses.

**Comment by ionelmc** [comment]
The licenses are the same. Am I missing something? :confused:

**Comment by ionelmc** [comment]
Also, for envs without virtualenv, there's always `pip install --user` (which doesn't need sudo).

**Comment by stevepeak** [comment]
Hey @ionelmc would you like to try #39 for your gcov builds. Do you have a repositories I can test against? Thanks!

ps: Now using `--user` for our python installer, works great! Thank you!

**Comment by ionelmc** [comment]
Sure, gonna try it on lazy-object-proxy (it has a C ext).

**Comment by ionelmc** [comment]
If you really want to do it yourself :grin: : https://github.com/ionelmc/python-lazy-object-proxy/blob/master/tox.ini#L86-L93

**Comment by thomasrockhu** [comment]
Hi @ionelmc, we meet again! I'm going to close this out due to the age of this request. If it's still an issue, let me know
